CMakeLists.txt

```diff
@@ -14,6 +16,50 @@ if (NOT DEFINED LLVM_SPIRV_BUILD_EXTERNAL)
   endif(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
 endif (NOT DEFINED LLVM_SPIRV_BUILD_EXTERNAL)
 
+# Download spirv.hpp from the official SPIRV-Headers repository.
+# One can skip this step by manually setting
+# LLVM_EXTERNAL_SPIRV_HEADERS_SOURCE_DIR path.
+if(NOT DEFINED LLVM_TOOL_SPIRV_HEADERS_BUILD AND
+   NOT DEFINED LLVM_EXTERNAL_SPIRV_HEADERS_SOURCE_DIR)
+  set(LLVM_EXTERNAL_SPIRV_HEADERS_SOURCE_DIR
+    "${CMAKE_CURRENT_BINARY_DIR}/SPIRV-Headers")
+  message(STATUS "SPIR-V Headers location is not specified. Will try to download
+          spirv.hpp from https://github.com/KhronosGroup/SPIRV-Headers into
+          ${LLVM_EXTERNAL_SPIRV_HEADERS_SOURCE_DIR}")
+  ExternalProject_Add(spirv-headers
+    GIT_REPOSITORY    https://github.com/KhronosGroup/SPIRV-Headers.git
+    GIT_TAG           origin/master
+    SOURCE_DIR        ${LLVM_EXTERNAL_SPIRV_HEADERS_SOURCE_DIR}
+    # Explicit specification of empty commands is needed to avoid "default"
+    # configure/build/install commands being run by the build system.
+    CONFIGURE_COMMAND ""
+    BUILD_COMMAND     ""
+    INSTALL_COMMAND   ""
+    STEP_TARGETS      build
+    COMMENT           "Downloading SPIR-V headers."
+    LOG_DOWNLOAD      1
+    LOG_UPDATE        1
+    LOG_BUILD         1
+  )
+else()
+  if(NOT DEFINED LLVM_EXTERNAL_SPIRV_HEADERS_SOURCE_DIR)
+    set(LLVM_EXTERNAL_SPIRV_HEADERS_SOURCE_DIR
+      "${CMAKE_SOURCE_DIR}/projects/SPIRV-Headers")
+  endif()
+  message(STATUS "Using SPIR-V Headers from
+          ${LLVM_EXTERNAL_SPIRV_HEADERS_SOURCE_DIR}")
+  ExternalProject_Add(spirv-headers
+    SOURCE_DIR ${LLVM_EXTERNAL_SPIRV_HEADERS_SOURCE_DIR}
+    DOWNLOAD_COMMAND  ""
+    # Explicit specification of empty commands is needed to avoid "default"
+    # configure/build/install commands being run by the build system.
+    CONFIGURE_COMMAND ""
+    BUILD_COMMAND     ""
+    INSTALL_COMMAND   ""
+    STEP_TARGETS      build
+  )
```

Comment:
Could you please explain the idea of this branch in more details? Because it seems like I got confused. Particular questions I have:
- why do we need `ExternalProject_Add` here without download command? Do we assume that there is like 100% guarantee that required files are there where we expect them?
- why do we need to do any extra actions for `projects/SPIRV-Headers` path at all? I though that it will be picked up by LLVM automatically

Reply:
> Why do we need ExternalProject_Add here without download command? Do we assume that there is like 100% guarantee that required files are there where we expect them?

What this command achieves is basically defining the `spirv-headers` target, which is used as a dependency for `LLVMSPIRVLib` from another branch. Even using the target_link_libraries approach from your [other comment](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/813/commits/92911948587adeac713ada01953080bc2dfdb47d#r662430738) would hardly resolve the need for defining this target - the download process still has to happen strictly before the `LLVMSPIRVLib` build/linkage.

> Why do we need to do any extra actions for projects/SPIRV-Headers path at all? I though that it will be picked up by LLVM automatically

If you're talking about the subbranch with `LLVM_TOOL_SPIRV_HEADERS_BUILD` defined, then unfortunately, LLVM does not define the path automatically. And there's still this point with the `spirv-headers` target requiring definition.

Reply:
Ok, let it be that it requires re-definition, but what if there are no headers in expected location? Should we add some kind of check (like to at least check if directory exists or if there is a `CMakeLists.txt`) and report an error if there are no headers in expected location?

Reply:
That's a fair point. Essentially, my logic has been that if `LLVM_TOOL_SPIRV_HEADERS_BUILD` has been defined, then the `CMakeLists.txt` presence has [already been checked](https://github.com/llvm/llvm-project/blob/main/llvm/cmake/modules/AddLLVM.cmake#L1350-L1354) by `add_llvm_subdirectory`. I agree that a check for the assumption of the `llvm/projects` location is still needed, and a check for `LLVM_SPIRV_BUILD_EXTERNAL OFF` also wouldn't hurt in this context.